### ISSUE TITLE
[v2 BUG]: ChempropMetric not working as expected with task_weights

### ISSUE THREAD

**Issue by AsaBarthMaron** [chemprop/chemprop#1111]
**Describe the bug**
When setting task_weights to anything other than [1, 1], the values produced by derived metrics and loss functions are incorrect if the mask differs between tasks. This is because of the way the mean loss is being calculated (dividing the total_loss by mask.sum()). On line [110 of metrics.py](https://github.com/chemprop/chemprop/blob/fa2d6dceb054ced6fe46191a628cfd664037d5e7/chemprop/nn/metrics.py#L110) rather than `self.num_samples += mask.sum()` it should be `(mask * self.task_weights).sum()` (Edit: will then need to be converted to long, e.g., (mask * self.task_weights).sum().long())

I.e., currently the metric is being normalized by the entire mask, not just the task-relevant mask

**Example(s)**
With equal task weights, but a mask that differs between tasks:
```
import torch
preds = torch.tensor([[0.2, 0.3], [0.4, 0.1], [0.2, 0.2]])
targets = torch.tensor([[0.3, 0.4], [0.4, 0.5], [1, 0.5]])
mask = torch.tensor([[1, 0], [1, 0], [0, 1]])
task_weights = torch.tensor([1, 1])

L = torch.abs(preds - targets) * task_weights * mask
print(L.sum() / mask.sum())
print(L.sum() / (mask * task_weights).sum())
```
The result is the same (0.1333 and 0.1333)

But with unequal task weights it differs:
```
import torch
preds = torch.tensor([[0.2, 0.3], [0.4, 0.1], [0.2, 0.2]])
targets = torch.tensor([[0.3, 0.4], [0.4, 0.5], [1, 0.5]])
mask = torch.tensor([[1, 0], [1, 0], [0, 1]])
task_weights = torch.tensor([1, 0])

L = torch.abs(preds - targets) * task_weights * mask
print(L.sum() / mask.sum())
print(L.sum() / (mask * task_weights).sum())
```
The result differs (0.0333 and 0.05)

**Environment**
- python 3.11.10
- package versions: Chemprop 2.1.0 official release, compiled from source
- OS - Ubuntu 22.04.5 LTS (SageMaker Distribution 2.1.0)

**Comment by KnathanM** [comment]
Hi @AsaBarthMaron, sorry it has taken me a while to get to this. And thank you for your examples of what you are thinking.

I think it would be more correct to calculate the metric for each task and then multiply each task's metric by its task weight before summing, but this isn't possible in our current implementation. We could change this in v2.2 (April/May), but I don't think it will make a big difference on model training. 

Possibly what you want to do is calculate the metrics for each task separately manually. This is what we do in the [CLI](https://github.com/chemprop/chemprop/blob/01d6ca4164858b2122b66204f8ca934e41efd51a/chemprop/cli/train.py#L1189-L1207).

Your question made me think hard about the weights we use when calculating the loss. Here is some background I wrote up to help me understand what is going on in case that is interesting to you. 

## Background
### Types of weights
There are three kinds of weights used by Chemprop metrics/loss functions:
* `weights` - This could also be called data point weights. It is used to weight the metric coming from all targets for a given data point. For example, if you include multiple representations of a single molecule, you could weight each of those `n` data points by `1/n` so that the model is less biased towards that single molecule. 
* `task_weights` - This is used to weight the metric from a target (task) for all data points. For example, if you have a lot of synthetic data for a property and a little bit of experimental data, you could try training a multi-task model to predict both the synthetic and experimental values, but weight the experimental task more. 
* `mask` - This isn't really a weight, but acts a bit like one. A Chemprop `MPNN` will [automatically](https://github.com/chemprop/chemprop/blob/3b8a4d4c08f5b88b2a289e321c77fc31a7865302/chemprop/models/model.py#L149) create a mask to mask out the metric coming from missing target/data point combinations. We set missing values to 0, calculate the metric vs. the predictions, and then ignore those metrics. 

### Metric vs Loss function
In Chemprop v1, metrics and loss functions were defined separately. In v2.0, duplicate code was reduced by having metrics inherit from their loss function counterparts where possible. This was then refined in v2.1 to follow the example of `torchmetrics` to have a loss function/metric pair only have one definition. In essence any metric can be used as a loss function if it is differentiable and if lower values are better. 

`task_weights` and (data point) `weights` are only intended to be used during training to help the model better focus on learning what we want it to. I recommend not using them when calculating metric values that you will then report, as the final value will depend on the arbitrary weights assigned. (I suppose there could be a metric with non arbitrary weights, but that isn't implemented in Chemprop.) For the rest of this, I'll refer to the output of the metrics as a loss. 

## Possible ways to use task weights
### Current
Our current loss calculation is 
```math
L_{total} = \frac{\sum_i \sum_j m_{ij} T_j d_i L_{ij}}{\sum_i \sum_j m_{ij}}
```
where $i$ is the data point index, $j$ is the task index, $m_{ij}$ is the mask, $T_j$ is the task weights, $d_i$ is the data point weight, and $L_{ij}$ is the loss from a data point/task combo.
 
### Task relevant mask
Your proposal is
```math
L_{total} = \frac{\sum_i \sum_j m_{ij} T_j d_i L_{ij}}{\sum_i \sum_j m_{ij} T_j}
```

I don't think this is correct though because the task weight will cancel out if a batch happens to only have data for one task (assuming the data point weights are all 1.0). This could happen if we are training a model to predict both a synthetic value and an experimental value and have much more synthetic data. Even if we weight the synthetic value task to be less important than the experimental value task, we would get a lot of batches like this (we don't have to sum over $j$ because there is only one task in the batch)
```math
L_{total} = \frac{\sum_i m_{i,syn} T_{syn} d_i L_{i,syn}}{\sum_i m_{i,syn} T_{syn}} = \frac{T_{syn} \sum_i m_{i,syn} d_i L_{i,syn}}{T_{syn} \sum_i m_{i,syn}} = \frac{\sum_i m_{i,syn} d_i L_{i,syn}}{\sum_i m_{i,syn}}
```

### Apply task weights after calculating loss
Instead we could calculate the loss for each task separately and then multiply by the task weights before summing. Actually this is what we do for `MCCLoss`. 
```math 
L_{total} = \sum_j T_j \frac{\sum_i m_{ij} d_i L_{ij}}{\sum_i m_{ij}}
```

But this would require an API change because `torchmetrics` metrics are initialized with "default" values for the internal states (which are used when resetting the metric). For `ChempropMetric`s these defaults are `torch.tensor(0.0)` for `total_loss` and `torch.tensor(0)` for `num_samples` (note that for `num_samples` it is a tensor with an integer which is why you needed to use `.long()`, this could be changed to a float if we wanted to multiply the mask by weights before summing to get `num_samples`). If we delayed using `task_weights` until after computing the loss for each task separately, then `total_loss` would need to be `torch.zeros(n_tasks)` and the user would need to specify how many tasks there are when initializing a metric. 

As an aside, we get around this problem in `MCCLoss` by appending a tensor of length `n_tasks` for every batch to a list and only combine batch values in `compute`. This isn't memory efficient, so I am hesitant to make all our metrics work like this. 

Maybe a better work around is to have `n_tasks` be a required argument to metrics so that the metric states defaults can be set to a tensor of the right shape. Maybe we could make it not required per se, but if `n_tasks` is `None` then we default to using a list to save the tensors from every batch. Either way, this would be a more major change and so we would implement it at the earliest in v2.2, which is slated for April.

But I'll also point out that the only difference between this and what we currently do is this divides the loss for each task by the number of samples for that task, while our current divides the loss for each task by the total number of samples for all tasks. I think both will work fine for training a model, so I am inclined to not make any changes. 


## What about data point weights?
Your question also made me wonder if we should include the data point weights in the number of samples? Something like `(mask * weights).sum()`. But similar to task weights, if there is a batch where all data points have the same weight, including the weight in the number of samples would cause the weight to just cancel out.

**Comment by KnathanM** [comment]
If there is more discussion about this topic in the future, feel free to reopen this issue.
